mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessageConverter.java

```diff
@@ -0,0 +1,138 @@
+
+    private final Clock clock;
+
+    DeletedMessageConverter(Clock clock) {
+        this.clock = clock;
+    }
+
+    DeletedMessage convert(DeletedMessageVaultHook.DeletedMessageMailboxContext deletedMessageMailboxContext, org.apache.james.mailbox.store.mail.model.Message message) throws IOException {
+        Preconditions.checkNotNull(deletedMessageMailboxContext);
+        Preconditions.checkNotNull(message);
+
+        Optional<Message> mimeMessage = parseMessage(message);
+
+        return DeletedMessage.builder()
+            .messageId(deletedMessageMailboxContext.getMessageId())
+            .originMailboxes(deletedMessageMailboxContext.getOwnerMailboxes())
+            .user(retrieveOwner(deletedMessageMailboxContext))
+            .deliveryDate(retrieveDeliveryDate(mimeMessage, message))
+            .deletionDate(ZonedDateTime.ofInstant(clock.instant(), ZoneOffset.UTC))
+            .sender(retrieveSender(mimeMessage))
+            .recipients(retrieveRecipients(mimeMessage))
+            .hasAttachment(message.getAttachments().iterator().hasNext())
+            .subject(mimeMessage.map(Message::getSubject))
+            .build();
+    }
+
+    private Optional<Message> parseMessage(org.apache.james.mailbox.store.mail.model.Message message) throws IOException {
+        DefaultMessageBuilder messageBuilder = new DefaultMessageBuilder();
+        messageBuilder.setMimeEntityConfig(MimeConfig.PERMISSIVE);
+        messageBuilder.setDecodeMonitor(DecodeMonitor.SILENT);
+        try {
+            return Optional.ofNullable(messageBuilder.parseMessage(message.getFullContent()));
+        } catch (MimeIOException e) {
+            LOGGER.warn("Can not parse the message {}", message.getMessageId(), e);
+            return Optional.empty();
+        }
+    }
+
+    private User retrieveOwner(DeletedMessageVaultHook.DeletedMessageMailboxContext deletedMessageMailboxContext) {
+        Preconditions.checkNotNull(deletedMessageMailboxContext.getOwner(), "Deleted mail is missing owner");
+        return deletedMessageMailboxContext.getOwner();
+    }
+
+    private ZonedDateTime retrieveDeliveryDate(Optional<Message> mimeMessage, org.apache.james.mailbox.store.mail.model.Message message) {
+        return mimeMessage.map(Message::getDate)
+            .map(Date::toInstant)
+            .map(instant -> ZonedDateTime.ofInstant(instant, ZoneOffset.UTC))
+            .orElse(ZonedDateTime.ofInstant(message.getInternalDate().toInstant(), ZoneOffset.UTC));
+    }
+
+    private MaybeSender retrieveSender(Optional<Message> mimeMessage) {
+        return mimeMessage
+            .map(Message::getSender)
+            .map(Mailbox::getAddress)
+            .map(MaybeSender::getMailSender)
+            .orElse(MaybeSender.nullSender());
+    }
+
+    private List<MailAddress> retrieveRecipients(Optional<Message> message) {
+        return StreamUtils.flatten(combineRecipients(message)
```

Comment:
I still think this part is really complex.

Reply:
The right place to handle this is in mime4j james utils

Well, a cross repo contrib is hard, and takes time / might include releases.

So I will uppost links here, but it might be cool to not await for the full process to finish.

Would you agree with this @aduprat  ?